doc/architecture.md

```diff
@@ -0,0 +1,206 @@
+      its child resources
+
+    - Rendering the resulting information in the form appropriate to the
+      resource and request method
+
+- The entry point to `DandiDav`'s functionality is its
+  [`handle_request()`][handle-request] method, which calls
+  `DandiDav::inner_handle_request()` and then performs error handling and
+  setting of universal response headers on the result.
+
+- [`DandiDav::inner_handle_request()`][] parses the request and passes the
+  results to either `DandiDav::get()` or `DandiDav::propfind()`, depending on
+  the request verb.
+
+- [`DandiDav::get()`][] and [`DandiDav::propfind()`][] both call
+  `DandiDav::get_resource_with_children()` to fetch information about the
+  requested resource and, if that resource is a collection, its immediate
+  children; as an exception, if a `PROPFIND` request was received with `Depth:
+  0`, `DandiDav::propfind()` will instead call `DandiDav::get_resource()`,
+  which does not obtain information about child resources.  `DandiDav::get()`
+  and `DandiDav::propfind()` then display the obtained information in different
+  ways: `DandiDav::get()` renders collections as HTML tables and
+  non-collections as redirects (or, for `dandiset.yaml`, serialized YAML),
+  while `DandiDav::propfind()` always returns a "multistatus" WebDAV XML
+  document.
+
+    - When rendering information about blob assets (either on their own or
+      within a parent resource), a choice must be made about what URL to
+      provide as the download URL.  The `contentUrl` metadata field of a blob
+      asset is assumed to contain two download URLs, one pointing to S3 and one
+      pointing to an Archive API endpoint that redirects to a signed version of
+      the S3 URL that sets a filename in the `Content-Disposition` header.  The
+      latter type of URL is generally more desirable, as it means that any
+      users downloading the URL via a browser will end up with a file with the
+      same filename as the asset, while a direct, unsigned S3 URL will be
+      downloaded as a file with the same filename as the S3 key, which is
+      usually the blob ID and not user-friendly.  Hence, by default, a `GET`
+      request to `dandidav` for a blob asset will be responded to with a
+      redirect to the Archive API download URL.  Unfortunately, certain WebDAV
+      clients (i.e., [davfs2](https://savannah.nongnu.org/bugs/?65376)) do not
+      support WebDAV servers that return redirects to other redirects, so the
+      `--prefer-s3-redirects` CLI option was added to `dandidav` to instead
+      make these `GET` requests redirect directly to the unsigned S3 URLs.
+
+        - Note that HTML listings of a collection's children will always link
+          blob assets to their Archive API download URLs, regardless of
+          `--prefer-s3-redirects`, as these listings are only returned for
+          `GET` requests to a collection, which WebDAV clients don't do.
```

Comment:
would tools like `wget` be accessing via `GET` or they would use webdav's PROPFIND to navigate?

Reply:
`wget` does not seem to support WebDAV, so it would be using `GET`.